test/metabase/db/connection_test.clj

```diff
@@ -0,0 +1,126 @@
+        do-in-transaction         (fn [thunk]
+                                    (try
+                                      (t2/with-transaction []
+                                        (thunk))
+                                      (catch Throwable e
+                                        (when-not (is-transaction-exception? e)
+                                          (throw e)))))]
+    (testing "inside transaction"
+      (do-in-transaction
+       (fn []
+         (create-user! user-1)
+         (is (user-exists? user-1))
+         (testing "inside nested transaction"
+           ;; do this on a separate thread to make sure we're not deadlocking, see
+           ;; https://github.com/seancorfield/next-jdbc/issues/244
+           (let [futur (future
+                         (do-in-transaction
+                          (fn []
+                            (create-user! user-2)
+                            (is (user-exists? user-2))
+                            (throw transaction-exception))))]
+             (is (not= ::timed-out
+                       (deref futur 1000 ::timed-out)))))
+         (testing "nested transaction aborted"
+           (is (user-exists? user-1))
+           (is (not (user-exists? user-2)))
+           (throw transaction-exception)))))
+
+    (testing "top-level transaction aborted"
+      (is (not (user-exists? user-1)))
+      (is (not (user-exists? user-2))))
+
+    (testing "make sure we set autocommit back after the transaction"
+      (t2/with-connection [^java.sql.Connection conn]
+        (t2/with-transaction [_t-conn conn]
+          ;; dummy op
+          (is (false? (.getAutoCommit conn))))
+        (is (true? (.getAutoCommit conn)))))
+
+    (testing "throw error when trying to create nested transaction when nested-transaction-rule=:prohibit"
+      (t2/with-connection [conn]
+        (t2/with-transaction [t-conn conn]
+          (is (thrown-with-msg?
+               clojure.lang.ExceptionInfo
+               #"Attempted to create nested transaction with :nested-transaction-rule set to :prohibit"
+               (t2/with-transaction [_ t-conn {:nested-transaction-rule :prohibit}]))))))
+
+    (testing "reuse transaction when creating nested transaction with nested-transaction-rule=:ignore"
+      (is (not (user-exists? user-1)))
+      (try
+        (t2/with-transaction []
+          (t2/with-transaction [t-conn]
+            (try
+              (t2/with-transaction [_ t-conn {:nested-transaction-rule :ignore}]
+                (create-user! user-1)
+                (is (user-exists? user-1))
+                (throw transaction-exception))
+              (catch Exception e
+                (when-not (is-transaction-exception? e)
+                  (throw e)))))
+          (is (user-exists? user-1))
+          (throw transaction-exception))
+        (catch Exception e
+          (when-not (is-transaction-exception? e)
+            (throw e))))
+      (is (not (user-exists? user-1))))
```

Comment:
it'd be much easier if we had some comments at each (is) here, this has 3 nested transactions and it's hard to figure out the intention of each test.